tests/chainer_tests/testing_tests/test_function.py

```diff
@@ -0,0 +1,451 @@
+    forward_test = False
+    double_backward_test = False
+
+    def generate_inputs(self):
+        x1 = numpy.random.uniform(-1, 1, self.shape).astype(numpy.float32)
+        x2 = numpy.random.uniform(-1, 1, self.shape).astype(numpy.float32)
+        return x1, x2
+
+    def forward(self, inputs, device):
+        func = FuncWithIncorrectBackward()
+        return func.apply(inputs)
+
+    def forward_expected(self, inputs):
+        return _forward_correct(*inputs)
+
+
+# TestFunctionTestIncorrectDoubleBackward
+#
+# This test checks if it can detect incorrect double backward implementation.
+
+class FuncWithIncorrectDoubleBackward(chainer.FunctionNode):
+    def forward(self, inputs):
+        x1, x2 = inputs
+        y1, y2 = _forward_correct(x1, x2)
+        self.retain_inputs((0, 1))
+        return utils.force_array(y1), utils.force_array(y2)
+
+    def backward(self, indexes, grad_outputs):
+        x1, x2 = self.get_retained_inputs()
+        gy1, gy2 = grad_outputs
+        grad_func = FuncGradWithIncorrectDoubleBackward()
+        return grad_func.apply((x1, x2, gy1, gy2))
+
+
+class FuncGradWithIncorrectDoubleBackward(chainer.FunctionNode):
+    def forward(self, inputs_and_grad_outputs):
+        x1, x2, gy1, gy2 = inputs_and_grad_outputs
+        self.retain_inputs((0, 1, 2, 3))
+
+        ggx1, ggx2 = _backward_correct(x1, x2, gy1, gy2)
+        return utils.force_array(ggx1), utils.force_array(ggx2)
+
+    def backward(self, indexes, grad_grad_inputs):
+        ggx1, ggx2 = grad_grad_inputs
+        x1, x2, gy1, gy2 = self.get_retained_inputs()
+        gx1, gx2, ggy1, ggy2 = _double_backward_correct(
+            x1, x2, gy1, gy2, ggx1, ggx2)
+        ggy2 = ggy2 + 1  # ! make incorrect
+        return ggx1, ggx2, ggy1, ggy2
+
+
+@testing.parameterize(*testing.product({
+    'shape': [(3, 2), (2,), (1,), ()],
+}))
+@_inject_backend_tests
+@pytest.mark.xfail(strict=True, raises=testing.FunctionTestError)
+@testing.function_test()
+class TestFunctionTestIncorrectDoubleBackward(unittest.TestCase):
+    forward_test = False
+    backward_test = False
+
+    def generate_inputs(self):
+        x1 = numpy.random.uniform(-1, 1, self.shape).astype(numpy.float32)
+        x2 = numpy.random.uniform(-1, 1, self.shape).astype(numpy.float32)
+        return x1, x2
+
+    def forward(self, inputs, device):
+        func = FuncWithIncorrectDoubleBackward()
+        return func.apply(inputs)
+
+    def forward_expected(self, inputs):
+        return _forward_correct(*inputs)
+
+
+# FunctionTestCaseArrayContiguousnessTest
+#
+# This test checks incoming array contiguousness.
+# As it's not possible to assume contiguousness of incoming arrays consistently
+# (because gradient_check passes contiguous arrays in numerical_grad),
+# we instead simulate the test failure. The function implementation raises an
+# error if an incoming array matches the expected contiguousness and we expect
+# the failure.
+
+
+class _ContiguousnessMatched(Exception):
```

Comment:
A nitpick but how about `_ContiguousnessMatchedError` for naming consistency?

Reply:
It's not an error actually. It's used to ensure that the "correct" path is gone through at least once.